### DISCUSSION TITLE
Unable to get proper vec4 value in uniform for frag shader

### DISCUSSION THREAD

**Discussion by sbrkopac** [vsg-dev/VulkanSceneGraph#1088] [Q&A] [answered]
[Here is an example program that shows my issue](https://gist.github.com/sbrkopac/18ae8880ea7cc33e889f197a2be0490d). I'm trying to add a "switch" to a uniform that is 1 or 0 and has a vec4 value associated with it. If it's 0 I want to use the vec4 as the color for the mesh rather than the texture.

I've been starting to use RenderDoc and when I look at a frame, I see the following:

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/1939712/c3f05793-2b30-45e2-a3dd-7c84591fb638)

Note that I'm expecting `vec4(1.0f, 1.0f, 0.0, 1.0f)`

When I look at the FS in RenderDoc:

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/1939712/9ec309b5-6cd4-4e29-ab53-b41594ae9218)

I believe I'm doing everything cleanly in the VSG side. Extra eyes would be great.

Thanks!

**Comment by rolandhill** [comment, marked as answer]
Hi Sam,

I'm wondering if you have fallen for the classic structure packing problem
(I certainly have). Shader structures are packed to 16 byte boundaries, so
your structure

layout(binding = 1) uniform Switch
{
    int useTexture;
    vec4 flatColor;
} switchData;

is probably getting 12 bytes extra padding between useTexture and
flatColor, which isn't accounted for when you load it on the CPU side. A
simple test would be to put an extra 3 x int dummy values in before
flatColor, or to switch useTexture and flatColor around so that flatColor
packs first (don't forget to do it in c++ code and in the shader).

Worth a try.

Cheers,
Roland

**Reply by sbrkopac** [reply]
Yup, that was exactly it. Thanks very much, Rolland!
